Implementation of pair_style uf3 and uf3/kk

monk-04 opened the pull request:
**Summary**

This pull request implements the [ultra-fast forcefield (uf3) machine learned interatomic potential](https://www.nature.com/articles/s41524-023-01092-7). Both the CPU and KOKKOS versions of uf3 are implemented.

<!--Briefly describe the new feature(s), enhancement(s), or bugfix(es) included in this pull request.-->

**Related Issue(s)**

None
<!--If this addresses an open GitHub issue for this project, please mention the issue number here, and describe the relation. Use the phrases `fixes #221` or `closes #135`, when you want an issue to be automatically closed when the pull request is merged-->

**Author(s)**

1. Ajinkya C Hire (University of Florida), email- ajinkya.hire@ufl.edu, ajinkyahire62@gmail.com
2. Hendrik Kraß (University of Constance)
3. Matthias Rupp (Luxembourg Institute of Science and Technology), email- mrupp@mrupp.info
4. Richard Hennig (University of Florida), email- rhennig@ufl.edu

<!--Please state name and affiliation of the author or authors that should be credited with the changes in this pull request. If this pull request adds new files to the distribution, please also provide a suitable "long-lived" e-mail address (ideally something that can outlive your institution's e-mail, in case you change jobs) for the *corresponding* author, i.e. the person the LAMMPS developers can contact directly with questions and requests related to maintenance and support of this contributed code.-->

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Will not break backward compatibility
<!--Please state whether any changes in the pull request will break backward compatibility for inputs, and - if yes - explain what has been changed and why-->

**Implementation Notes**

The correctness of the implementation was verified by comparing the lammps computed energy and forces with that calculated from the python version of uf3 found [here](https://github.com/uf3/uf3). The current version computes two-body and three-body interactions, contributions from one-body are ignored.
<!--Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected-->

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

<!--**Further Information, Files, and Links**

Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

monk-04 committed:
Added ml-uf3 src files

monk-04 committed:
Added src files for uf3/kk

monk-04 committed:
Added ml-uf3 to cmake and make files

monk-04 committed:
Added documentation for ml-uf3

monk-04 committed:
Added ml-uf3 potential files

monk-04 committed:
Added ml-uf3 unittest

monk-04 committed:
Fixed typos in the UF3 equation

monk-04 committed:
Fixed lammps developer email-id and contributing authors section. Removed some old comments

monk-04 committed:
Removed LAMMPS errordocs

monk-04 committed:
fixed trailing whitespaces

monk-04 committed:
Added uf3 details to Commands_pair, Packages_details and Packages_list

monk-04 committed:
Added uf3 examples to the examples directory

monk-04 committed:
Removed trailing whitespaces

monk-04 committed:
Added example potential files for W

akohlmey committed:
add ML-UF3 to compatible CMake preset files

akohlmey committed:
update .gitignore

akohlmey committed:
update code owners list

akohlmey committed:
move examples

akohlmey committed:
reformat and fix spelling and related issues

akohlmey committed:
update ML-UF3 examples

akohlmey committed:
some style cleanup and simplified pair_style and pair_coeff input

srtee commented on src/ML-UF3/pair_uf3.cpp:
```diff
@@ -0,0 +1,1313 @@
+        del_rki[1] = x[k][1] - ytmp;
+        del_rki[2] = x[k][2] - ztmp;
+        rik = sqrt(
+            ((del_rki[0] * del_rki[0]) + (del_rki[1] * del_rki[1]) + (del_rki[2] * del_rki[2])));
```
Lots of unnecessary rij recalculations here -- see comment at line 1015

srtee commented on src/ML-UF3/pair_uf3.cpp:
```diff
@@ -0,0 +1,1313 @@
+      if (rsq < cutsq[itype][jtype]) {
+        rij = sqrt(rsq);
+
+        if (pot_3b) {
+          if (rij <= cut_3b_list[itype][jtype]) {
+            neighshort[numshort] = j;
+            if (numshort >= maxshort - 1) {
+              maxshort += maxshort / 2;
+              memory->grow(neighshort, maxshort, "pair:neighshort");
+            }
+            numshort = numshort + 1;
+          }
+        }
```
from what I understand, this creates an internal "shorter-neighbor list" within which to evaluate 3body interactions. To me a simpler approach would be:
1. something like `double * rij_list = new double[inum]` (but create and manage the memory in a broader context for better predictability and performance)
2. after calculating each `rij`, store it: `rij_list[jj] = rij`
3. you now have the distances on hand for 3-body interactions later on

An extra `inum`-worth of doubles isn't a prohibitive amount of memory -- 1000 doubles is 8 KB.

srtee replied:
You could even afford to store the delx, dely, and delz doubles in their own arrays as well.

monk-04 replied:
I agree storing an array of rij will reduce the number of computations later on.
But I also need to know the type of atoms in the 3b interaction evaluation. So I need to keep the current `neighshort`.

Also, in practice, the 3b cutoffs are always shorter than 2-body cutoffs (usually by a couple of angstroms). Do you still think I need to make the `rij_list` of the size of `inum`?

srtee replied:
My main worry is a bunch of unpredictable memory realllocations happening inside the inner loop of a force calculation. The `if(pot3b) ...` code is being triggered `nlocal * jnum` times and all it takes is one processor needing to do a `grow` to imbalance that timestep's force calculation.

Types are straightforward to track: if `jj` runs from 0 to `jnum`, the distance `rij_list[jj]` belongs to particle `jlist[jj]` which has particle type `type[jlist[jj]]`.

monk-04 replied:
> Types are straightforward to track: if `jj` runs from 0 to `jnum`, the distance `rij_list[jj]` belongs to particle `jlist[jj]` which has particle `type type[jlist[jj]]`.

I am bit confused.
`jnum` is decided by the cutoff of 3body interaction and, in practice, is always less than 2body.

In `neighshort`, I am storing `j` ie `jlist[jj]` during the 2body loop. Not every `j` in 2body loop forms a valid 3body triangle. 

So, if my understanding of your suggestion is correct, you are suggesting to set `jnum` of 3body loop equal to `jnum` of 3body loop(?). If that is done, a lot of iterations in the 3body loop will do nothing because of the distance  cutoff checks. 

I apologize if I am not understanding your suggestion correctly.

srtee commented on src/ML-UF3/pair_uf3.cpp:
```diff
@@ -0,0 +1,1313 @@
+
+          std::string key = std::to_string(i) + std::to_string(j) + std::to_string(k);
+
+          for (int l=0; l < n3b_coeff_matrix[key].size(); l++){
+            for (int m=0; m < n3b_coeff_matrix[key][l].size(); m++){
+              bytes += (double)2*n3b_coeff_matrix[key][l][m].size()*sizeof(double);
```
I don't see that n3b_coeff_matrix is used outside the initialization functions -- clearing it after initialization would save some memory (probably not much?).

monk-04 replied:
I am passing n3b_coeff_matrix[key] by reference to 3b bspline object (triplet_bspline UFBS3b), which then uses it to calculate energies and forces

srtee commented on src/ML-UF3/uf3_pair_bspline.cpp:
```diff
@@ -0,0 +1,157 @@
+  ret_val[1] += dnbspline_bases[knot_affect_start].eval2(rsq, r);
+
+  return ret_val;
+}
```
Instead of returning a pointer-to-double you can return a force double and modify the energy by reference. So the function signature would be

`double uf3_pair_bspline::eval(double r, double &eng)`

and your evaluations would proceed as

```
... eng = 0
    eng += \* first energy component *\
    eng += \* second energy component *\
...
    f = 0
    f += \* first force component *\
    f += \* second force component *\
... return f
```

I am not sure how much this changes performance but it will make code more readable for future maintainers.

srtee commented on src/ML-UF3/pair_uf3.cpp:
```diff
@@ -0,0 +1,1313 @@
+          }
+        }
+
+        double *pair_eval = UFBS2b[itype][jtype].eval(rij);
+
+        fpair = -1 * pair_eval[1] / rij;
+
+        fx = delx * fpair;
+        fy = dely * fpair;
+        fz = delz * fpair;
+
+        f[i][0] += fx;
+        f[i][1] += fy;
+        f[i][2] += fz;
+        f[j][0] -= fx;
+        f[j][1] -= fy;
+        f[j][2] -= fz;
+
+        if (eflag) evdwl = pair_eval[0];
```
See comment on the `eval` function -- it may be more readable to modify energy by reference. So the function signature would be

`double CLASSNAME::eval(double r, double &eng)`

and it could be called as

`fpair = -1 * UFBS2b[itype][jtype].eval(rij, evdwl) / rij;`

This has minor readability impact here -- later on in the 3-body functions, it is extremely useful to have the variable names in both the function definition in the spline CPPs and in the function call here, instead of having to separately document what each index in a returned double pointer means.

monk-04 replied:
Should I also pass eflag to the eval function? Something like `UFBS2b[itype][jtype].eval(rij, evdwl, eflag)`? This will save some computations as energy evaluation is not needed on every timestep?

srtee replied:
If the energy calculation is costly enough, then yes, you can pass eflag and let `eval` save some time. But you would know better whether that's worth it.

As a style thing I'd prefer to keep the reference doubles together after the value doubles:

```
double CLASSNAME::eval(
  double rij,
  int eflag,
  double& evdwl // return energy by ref
){ ...
```

akohlmey replied:
> Should I also pass eflag to the eval function? Something like `UFBS2b[itype][jtype].eval(rij, evdwl, eflag)`? This will save some computations as energy evaluation is not needed on every timestep?

since eflag doesn't change, the most efficient way to handle this would be to make eval a templated function and eflag a template argument. That could hoist the if statement to the outermost level and the if statement will be optimized away since you either call `eval<0>()` or `eval<1>()`. The compiler will then optimize the "if" away since the condition is a compile time constant. Check out examples in OPT or OPENMP.

srtee commented on src/ML-UF3/pair_uf3.cpp:
```diff
@@ -0,0 +1,1313 @@
+            f[k][1] += Fk[1];
+            f[k][2] += Fk[2];
+
+            if (eflag) evdwl = *triangle_eval;
```
This is where a call-by-reference method can be much clearer. For consistency with the 2body call I'd make `eval` still return `evdwl` by reference, so this `eval` has to be `void` (there isn't a natural particular double to return). But you could call it as:

```
UFBS3b[itype][jtype][ktype].eval(rij, rik, rjk,
Fmag_ij, Fmag_ik, Fmag_jk, evdwl)
```

and voila! Each of your return variables are now named, not numbered!

akohlmey commented on cmake/CMakeLists.txt:
There should be updates for the preset files. Included in my PR to your branch.

akohlmey committed:
examples folder was moved

akohlmey committed:
manybody and single flag need to be changed from the default when 2-body/3-body is selected

akohlmey commented on potentials/A_A_A.uf3_pot:
What element does A stand for?

monk-04 replied:
The potential files were modeled after Niobium UF3 potential files. I will replace them with the real files.

akohlmey commented on doc/src/pair_uf3.rst:
```diff
@@ -0,0 +1,184 @@
+   pair_coeff 3b 2 1 2 B_A_B
+   pair_coeff 3b 2 2 2 B_B_B
+
+If a value of "2" is specified in the :code:`pair_style uf3` command, only the two-body potential files are needed. For 3-body interaction the first atom type is the central atom. We recommend using the :code:`generate_uf3_lammps_pots.py` script (found `here <https://github.com/uf3/uf3/tree/master/lammps_plugin/scripts>`_) for generating the UF3 LAMMPS potential files from the UF3 JSON potentials ...
```
If possible this script should be added to the `tools` tree in `tools/uf3`.

monk-04 replied:
The python script depends on the `uf3` python package and is not made to work out of the box. Should I still add it?

akohlmey replied:
No that would be overkill.

akohlmey commented on doc/src/pair_uf3.rst:
```diff
@@ -0,0 +1,184 @@
+
+.. code-block:: LAMMPS
+
+   #UF3 POT
```
The potential files should have metadata tags that indicate the source, a reference, the data, and in particular can indicate the units, so that LAMMPS will refused to read potential files created for incompatible units.
You are already using `utils::open_potential()` so this check will be automatic, provided the metadata tags are present.

monk-04 replied:
Yup, working on it

akohlmey commented on src/ML-UF3/pair_uf3.cpp:
```diff
@@ -0,0 +1,1313 @@
+
+void PairUF3::uf3_read_pot_file(int itype, int jtype, char *potf_name)
+{
+  utils::logmesg(lmp, "UF3: {} file should contain UF3 potential for {} {}\n", \
```
This will print one line per MPI ranks. Not a good idea. Since the number of pair_coeff lines for systems with multiple elements can be large, it is probably best to drop this output altogether.

akohlmey commented on src/ML-UF3/pair_uf3.cpp:
```diff
@@ -0,0 +1,1313 @@
+  utils::logmesg(lmp, "UF3: {} file should contain UF3 potential for {} {}\n", \
+          potf_name, itype, jtype);
+
+  if (!platform::file_is_readable(potf_name))
```
This is redundant you can check for the file pointer being NULL below.

akohlmey committed:
update more files and docs for .uf3 potential file extension

also remove redundant files

akohlmey committed:
some formatting and logic cleanup.

akohlmey commented on src/ML-UF3/pair_uf3.cpp:
```diff
@@ -0,0 +1,1313 @@
+    error->all(FLERR, "UF3: {} file is not readable", potf_name);
+
+  FILE *fp;
+  fp = utils::open_potential(potf_name, lmp, nullptr);
```
these two lines could be merged and the fest for failure should be added below.

akohlmey commented on src/ML-UF3/pair_uf3.cpp:
```diff
@@ -0,0 +1,1313 @@
+  Tokenizer file_header(temp_line);
+
+  if (file_header.count() != 2)
+    error->all(FLERR, "UF3: Expected only two words on 1st line of {} but found \n\
```
The trailing backslash is rather unusual. Is fixed in my PR branch. also, there don't need to be so many newlines.

akohlmey reviewed (changes requested):
This is it for a first round of review. I didn't change most of the unprotected utils::logmesg() calls. In my opinion most should just be removed. They look like debugging output.

One more general comment. It is generally considered a bad idea to read the same file from all the MPI ranks. This can create huge problems with jobs using many MPI processes. The more common procedure is to read everything on rank 0 only and then use MPI communication to broadcast the information to all other ranks.

monk-04 committed:
Merge pull request #1 from akohlmey/ml-uf3

Updates for ml-uf3 pull request

monk-04 commented:
@akohlmey Thank you for your comments and edits.

> It is generally considered a bad idea to read the same file from all the MPI ranks

I will modify the code so that only rank 0 will read the potential file.


>utils::logmesg() 

Many of the logmesg were put in the code at the request of early users of the `pair_style`. I will think about this point a bit more.

- [ ] Pimplify pair_uf3.cpp
- [ ] Read UF3 potential files on rank 0 and use MPI communicate to other processors

Currently the unittest is failing because of the following error-

`LAMMPS Error: ERROR on proc 0: Cannot open UF3 potential file A_A.uf3 (src/ML-UF3/pair_uf3.cpp:215)
Last command: pair_coeff * * A_A.uf3`

I think this is because the `A_A.uf3` and `A_A_A.uf3` potential files were moved to the `potentials` directory.

akohlmey committed:
update example logs

akohlmey committed:
remove dead code

akohlmey committed:
add system error message with failure to open potential

akohlmey commented:
> I think this is because the `A_A.uf3` and `A_A_A.uf3` potential files were moved to the `potentials` directory.

No. It is because you didn't close the files. The unit test runs will execute multiple short runs and after several of them, you run out of available file descriptors. That is when the open failed. I'll improve the error message.

akohlmey committed:
apply clang-format

akohlmey committed:
call utils::logmesg() only on rank 0, use c++ string comparisons, remove debug comments

akohlmey committed:
don't use pow() function for simple square

akohlmey committed:
remove num_of_elements class variable, just use local copy of atom->ntypes

akohlmey committed:
initialize all class pointers to null, delete all allocated storage

akohlmey committed:
drop log messages

akohlmey committed:
replace pow(x,2) and pow(x,3) with square() and cube()

akohlmey commented:
I've submitted https://github.com/uf3/lammps/pull/2 with more changes including the file handle leak fix.

akohlmey commented:
> * Read UF3 potential files on rank 0 and use MPI communicate to other processors

Here are more things to consider:
- add metadata to UF3 potential files so that units can be checked
- use a map so that unique potential files and its parameters need to be read only once. For example in the unit test input you need to read the same A_A_A.uf3 potential file 512 times and store 512 copies of the same parameters. with a map only one read and one copy would be needed. Check out other manybody potentials like Tersoff where they have a elem3param list
- consider moving away from the vector of vector (of vector) data storage toward multi-dimensional arrays created with the memory class. These are much easier to handle for broadcasting data. You just need to broadcast the dimensions (if they are not already known), destroy then create the arrays and broadcast the data in fortran style with a single  MPI_Bcast() of &array[0][0] or similar since the data holding block is consecutive like in fortran. There are examples for that all over the LAMMPS code.

akohlmey commented:
> I've submitted [uf3#2](https://github.com/uf3/lammps/pull/2) with more changes including the file handle leak fix.

You need to merge this pull request to have it pass the tests here.

monk-04 commented:
I am running some numerical accuracy (comparison with Python) and speed tests on our systems. They should be over in some time.

monk-04 committed:
Merge pull request #2 from akohlmey/ml-uf3

More updates to PR 4110

monk-04 commented:
I have merged the PR.

Next I will work on the following things-

- [x] Pimplify pair_uf3.cpp
- [x] Add `UNITS`, `AUTHOR`, `DATE` and `CITATION` metadata
- [x] Read all UF3 potentials from a single file
- [x] Use arrays from memory class instead of std::vectors
- [x] Read UF3 potential files on rank 0 and use MPI to communicate to other processors
- [ ] ~~Modify `eval` of 2body bspline classes to take  `rij`, `evdwl`, and `eflag` as input and return only `fpair`~~
- [ ] ~~Modify `eval` of 3body bspline classes to also take `Fmag_ij`, `Fmag_ik`, `Fmag_jk`, `evdwl`, and `eflag` as input~~

monk-04 committed:
Pimplifying the code Attempt-1

monk-04 committed:
Merge branch 'lammps:develop' into ml-uf3

monk-04 committed:
Removed commented out code

monk-04 committed:
Added code to check if 'UNITS:' metadata is present in the pot file or not

monk-04 committed:
Replaced the A_A.uf3 and A_A_A.uf3 with uf3 lammps pot files for Nb

monk-04 committed:
Updated A_A.uf3 to Nb_Nb.uf3 and A_A_A.uf3 to Nb_Nb_Nb.uf3

monk-04 committed:
Removed old pot files

monk-04 committed:
Updated documentation about METADATA in the uf3 lammps pot files

monk-04 committed:
Fixed trailing whitespace

monk-04 committed:
Merge branch 'lammps:develop' into ml-uf3

monk-04 committed:
Added uf3_read_unified_pot_file() to read single potential file on rank 0 and communicate() to broadcast the data to other ranks

Added one_ceoff

monk-04 committed:
Added new constructor functions in uf3_pair_bspline and uf3_triplet_bspline to construct std::vectors of knots and coefficients rom memory block

monk-04 committed:
Updated pair_coeff in unittest to read only one potential file; added Nb.uf3 unified potential file

monk-04 committed:
Updated the code as we are no longer using std::vector for knots and coefficients

monk-04 committed:
Removed reference to std::vector knots and coefficients; commented out getter functions

monk-04 committed:
Updated the documentation about UF3 LAMMPS potential file

monk-04 committed:
Merge pull request #4 from monk-04/ml-uf3

Updates to pair_uf3- Reading potential from single file, arrays from memory class, MPI to communicate the arrays

monk-04 committed:
Fixed trailing whitespaces

monk-04 committed:
Merge branch 'lammps:develop' into ml-uf3

monk-04 committed:
Made 'constants' variable public

monk-04 committed:
Removed UFBS2b and UFBS3b std::vector objects.

These objects held the bspline basis set object (uf3_pair_bspline and uf3_triplet_bspline) for 2- and 3-body interaction at UFBS2b[itype][jtype] and UFBS3b[itype][jtype][ktype].
These std:vectors were removed. New arrays (cached_constants_2b, cached_constants_2b_deri, cached_constants_3b, cached_constants_3b_deri) were added that holds the cached coefficients of the bspline basis set. The UF3Impl PIMPLE was also removed as it is not longer needed. The memory_usage function needs to updated

monk-04 committed:
Removed trailing whitespace

monk-04 committed:
constants was changed from std::vector to fixed length array

monk-04 committed:
Removed uf3_pair_bspline and uf3_triplet_bspline

monk-04 committed:
Commented out the uniform knot spacing logic as for more than 1 processors was getting some weird errors.
Deleted commented (dead) code.
Updated the memory_usage function.
Reordered some functions to refelect the calling order

monk-04 committed:
Removed unused variables and reordered the code

monk-04 committed:
Fixed trailing whitespace

monk-04 committed:
Removed dead code, fixed trailing whitespace

monk-04 committed:
Merge pull request #5 from monk-04/ml-uf3

Remove std::vector completely

monk-04 committed:
Merge branch 'lammps:develop' into ml-uf3

monk-04 committed:
Fixed bug with uniform knot spacing

monk-04 committed:
Initialize setflag, n2b_coeff_array_size, n2b_knots_array_size, n3b_coeff_array_size, n3b_knots_array_size arrays to 0; fixed memory leaks; removed some dead code

monk-04 committed:
Removed trailing whitespace

monk-04 committed:
Merge pull request #6 from monk-04/ml-uf3

Bug Fixes- uniform knot spacing, memory leaks, array initialization

monk-04 commented:
@akohlmey Thank you for your suggestion about moving away from std::vectors. It forced me to re-write certain functions. Now arrays are used to store constants of bspline basis set instead of storing uf3_pair_bspline and uf3_triplet_bspline objects (at UFBS2b std::vector\<std::vector\> and UFBS3b std::vector\<std::vector\<std::vector\>\>). Also, I evaluate the energy contribution only when 'eflag' is true instead of every time step.

All these changes led to about 30%-35% speed up for the CPU version.

stanmoore1 commented:
@monk-04 is this PR ready for review or are more changes coming?

monk-04 commented:
@stanmoore1 I was waiting for PR #4146 so that the MacOS unit test could pass.
I see that it has been merged and will merge lammps:develop into ml-uf3.

Other than that, I don't have any more changes in mind.

The PR is ready for review.

monk-04 committed:
Merge branch 'lammps:develop' into ml-uf3

akohlmey committed:
Merge branch 'develop' into ml-uf3

akohlmey committed:
fix spelling

akohlmey committed:
remove unused variables

akohlmey committed:
update ML-UF3 example

akohlmey commented:
@monk-04 I've submitted a pull request with a few minor changes and replaced the example.

monk-04 committed:
Merge pull request #7 from akohlmey/ml-uf3

Update example with new syntax

stanmoore1 commented:
@monk-04 I took a look at the Kokkos code (thanks for adding that btw), it looks pretty good. I have a few cleanup changes that I will submit as a PR to your fork. The Kokkos version does give correct numerics, but doesn't saturate the GPU with only a couple hundred atoms in the UF3 example. If you plan to run with small atom counts/GPU, it could be beneficial to thread over both atoms and neighbors as done here to expose more parallelism: https://github.com/lammps/lammps/blob/develop/src/KOKKOS/pair_kokkos.h#L604-L780. Otherwise you will need ~500k or more atoms/GPU to saturate, based on other potentials I've run.

monk-04 commented:
@stanmoore1 Thank you for looking through the code.

The current code needs a very large system size to saturate the GPU-
![image](https://github.com/lammps/lammps/assets/48733363/f2d58fac-1df0-46ab-86e5-0a9470430bc1)
(Run on 1 A100)


I will look in to the code you suggested, and will try to implement it.

akohlmey commented:
@monk-04 we're merging this pull request now, so it will be in the next feature release and thus also the stable release. You can create a new pull request with optimizations later when you have them ready. If we would wait for them, chances are that your contribution would have to be postponed until after the stable release.

monk-04 commented:
@akohlmey That works for me. I am working on my thesis, so any more optimizations will take a bit more time from my side. Thank You.

akohlmey commented:
@monk-04 FYI, we may have to pull the KOKKOS version of your pair style from the distribution until it is corrected. It is missing the "host" version of the implementation and fails compilation under certain circumstances.  E.g. I just ran into issues trying to compile with OpenMP and HIP (for an AMD GPU).

Ping me on the LAMMPS developer slack for more details.

monk-04 commented:
Ok. I will message you on Slack.
